docs/arch/adr-03.md

```diff
@@ -0,0 +1,112 @@
+* access permission features
+* Speed of accessing and iterating through files (e.g. getting all files in a given directory)
+
+### Terminology:
+
+**Blob:** acronym for “Binary Large Object”  
+**Container:** synonym for ”S3 Bucket”  
+**Shared Access Signature:** similar functionality as “S3 Presigned URLs”
+
+## Considered Options
+
+* Blob Storage
+* File Services
+
+> Azure Files and Azure Blob Storage both offer ways to store large amounts of data in the cloud, but they are useful for slightly different purposes.
+>
+> Azure Blob Storage is useful for massive-scale, cloud-native applications that need to store unstructured data. To maximize performance and scale, Azure Blob Storage is a simpler storage abstraction than a true file system. You can access Azure Blob Storage only through REST-based client libraries (or directly through the REST-based protocol).
+>
+> Azure Files is specifically a file system. Azure Files has all the file abstracts that you know and love from years of working with on-premises operating systems. Like Azure Blob Storage, Azure Files offers a REST interface and REST-based client libraries. Unlike Azure Blob Storage, Azure Files offers SMB access to Azure file shares. By using SMB, you can mount an Azure file share directly on Windows, Linux, or macOS, either on-premises or in cloud VMs, without writing any code or attaching any special drivers to the file system. You also can cache Azure file shares on on-premises file servers by using Azure File Sync for quick access, close to where the data is used.
+
+Both options allow for specifying read-only or write-only permissions on folders or files using [Shared Access Signatures](https://docs.microsoft.com/en-us/rest/api/storageservices/create-user-delegation-sas).
+
+### Option 1: Azure Blob Storage
+
+#### Summary 
+Blob Storage is optimized for storing unstructured data: e.g. information that doesn't reside in a traditional row-column database. Blobs come in [three different types](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs): block blobs, append blobs, and page blobs.
+
+##### Block Blobs
+ A block blob consists of one or more blocks, each each of which is identified by a block ID. 
+
+By default, storage clients limit block size to 128 MB, such that when a block blob upload is more than 128 MB, the file will be broken into several blocks. This maximum block size can be updated via the SingleBlobUploadThresholdInBytes property. 
+
+When uploading additional blocks to a blob, the additional blocks are associated with the specified blob, but do not become part of the blob until you commit a list of block IDs that include the newly uploaded blocks.
+
+We should note that the Azure Blob Storage gem only exposes the functionality for working with block blobs - to work with any other type of blob, we would have to construct HTTP requests by hand.
```

Comment:
I think we treat the file upload as a blob and the gem does all the block management etc. Can you clarify if the gem abstracts away the interaction with individual blocks and managing the commits?

Reply:
Seems like it. When you call `#create_block_blob`, the source (https://github.com/Azure/azure-storage-ruby/blob/master/blob/lib/azure/storage/blob/block.rb) calls either `  #create_block_blob_single_put` or `#create_block_blob_multiple_put` (depending on the threshold size, default 1MB) to put either a single blob or a block (of multiple blobs). Only the multiple put explicitly commits, but since it does I think we can assume that it would happen either way.

Reply:
Yep exactly! Like @zwolf said, you pass a file to the gem function `#create_block_blob` and it abstracts away the creation of the actual blocks. @camallen , feel free to check out usage section on using the `azure-storage-blob` gem (https://github.com/azure/azure-storage-ruby/tree/master/blob#usage)